File: src/Java/algorithms/linkedlist/SortList.java
package algorithms.linkedlist; 

import java.util.*; 
import lib.ListNode; 

public class SortList {
    /* Function: Sort linked list in O(nlogn). 
     * Solution: Merge sort. 
     */ 
    
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head; 
        }
        
        return sortHelper(head); 
    }
    
    private ListNode sortHelper(ListNode head) {
        if (head == null || head.next == null) {
            return head; 
        }
        
        ListNode fast = head.next, slow = head; 
        while (fast != null && fast.next != null) {
            fast = fast.next.next; 
            slow = slow.next; 
        }
        
        ListNode right = sortHelper(slow.next); 
        slow.next = null; 
        ListNode left = sortHelper(head); 
        return merge(left, right); 
    }
    
    private ListNode merge(ListNode left, ListNode right) {
        ListNode dummy = new ListNode(0), cur = dummy; 
        while (left != null || right != null) {
            if (right == null || (left != null && left.val < right.val)) {
                cur.next = left; 
                left = left.next; 
            } else {
                cur.next = right; 
                right = right.next; 
            }
            
            cur = cur.next; 
        }
        
        return dummy.next; 
    }
} 

File: src/CPP/src/leetcode/SortList.cpp
/*
 * [Source] https://leetcode.com/problems/sort-list/
 * [Difficulty]: Medium
 * [Tag]: Linked List
 * [Tag]: Sort
 */

#include <iostream>

using namespace std;

// [Solution]: Count the length of the list first. Then break the list into half use this number. Run quick sort.
// [Corner Case]:
class Solution {
};

// [Solution]: Use two pointers
/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/linkedlist/SortList.java
 */

/* Java solution
public class Solution {
    public ListNode sortList(ListNode head) {
		if(head==null || head.next==null) return head;
		ListNode end=head;
		ListNode mid=head;
		ListNode current=head;
		while(end!=null && end.next!=null){
			end=end.next.next;
			current=mid;
			mid=mid.next;
		}
		current.next=null;
		ListNode first=sortList(head);
		ListNode second=sortList(mid);
		head=mergeList(first, second);
		return head;
    }
	public ListNode mergeList(ListNode first, ListNode second){
		ListNode dummy=new ListNode(0);
		ListNode current=dummy;
		while(first!=null && second!=null){
			if(first.val<second.val){
				current.next=first;
				current=first;
				first=first.next;
			}else{
				current.next=second;
				current=second;
				second=second.next;
			}
		}
		if(first==null) current.next=second;
		else current.next=first;
		return dummy.next;
	}
    
}
 */

int main() {
    Solution sol;

    return 0;
}
